### ISSUE TITLE
Allow the user to load images within SSR (without bot header)

### ISSUE THREAD

**Issue by intellix** [tjoskar/ng-lazyload-image#414]
So we're using SSR and am currently trying to optimise the loading experience for the user. Currently all of our images are using `[lazyLoad]="..."` and it seems to work perfectly in the browser. On the server the images are never rendered in the universal delivered response. 

We'd like that when universal sends the response, the image.src properties are set and we can http/2 push images above the fold (most optimal) along with index.html. 

Currently the universal response is around 700kb and we have to wait until JS has loaded before the images appear in front of the user and they're staring at a blank areas until then.

The next best thing would be to have lazyLoading on the browser and to ignore it entirely on SSR (for the best initial load experience of the first render).

As it stands I can't get it to work at all, even when returning:

```
isBot() { return true }
```

I think because the call to lazy-load is run outside of angular and the page is considered done and delivers without them:
https://github.com/tjoskar/ng-lazyload-image/blob/41b537cb730c35932417271ef1aa26a6e40bb539/src/lazyload-image.directive.ts#L58-L65

**Comment by tjoskar** [comment]
Thanks for the repo!

> We'd like that when universal sends the response, the image.src properties are set and we can http/2 push images above the fold (most optimal) along with index.html.

This would be super nice but in order to achieve this, I think we need to render the page in a browser on the server (like puppeteer or electron or similar), it will otherwise be impossible (or very hard at least) to determine the position (x and y) of an image. I think this is outside the scope of this project but I think it would be nice to actually do it.

> The next best thing would be to have lazyLoading on the browser and to ignore it entirely on SSR

You mean to not lazyload the images on the first-page view but then lazyload the images on the rest pages?

And yes, there is a bug right now when you can't overwrite the `isBot` function under `ssr`. 

https://github.com/tjoskar/ng-lazyload-image/blob/41b537cb730c35932417271ef1aa26a6e40bb539/src/hooks-factory.ts#L8-L14

Line 11 should be something like:
```ts
return Object.assign(ssrPreset, { isBot });
```

So one solution for you could be to fix the bug above and then pass `isBot: () => true` but you probably don't want that in the browser, since non of the image will be lazyloaded, ever.

An other solution would be to pass the `platformId` to `isBot` (as you sugested in an other comment). In that way can you create following function: `isBot: platformId => isPlatformServer(platformId) ? true : intersectionObserverPreset.isBot(navigator)`. And I really liked your idea of passing `platformId` to the function. 

I may have the time tomorrow to create a PR that you can test otherwise are PRs always welcome :)

**Comment by intellix** [comment]
I had a go at the `Object.assign(ssrPreset, { isBot})` by changing it directly in my server bundle but didn't quite get it to work. I'm under the impression that running it outside of Angular makes it async and Universal regards it as ready/finished but I'm not entirely sure without more testing

**Comment by tjoskar** [comment]
I don't think the problem is `runOutsideAngular` but there is some problem with `forRoot` with AOT that I don't understand.

This doesn't work: 

```ts
LazyLoadImageModule.forRoot({
  isBot() { return true; }
}),
```
No error but `forRoot` is never called and `isBot` is not defined.

This doesn't work either:
```ts
LazyLoadImageModule.forRoot({ isBot: () => { return true; } }),
// nor
const isBot = () => { return true; }
LazyLoadImageModule.forRoot({ isBot }),
```
With the error: `Function expressions are not supported in decorators in 'isBot'`

This doesn't work either:
```ts
function isBot() {
  return true;
}

LazyLoadImageModule.forRoot({ isBot }),
```
With the error: `Reference to a non-exported function.`

This however works:
```ts
export function isBot() {
  return true;
}

LazyLoadImageModule.forRoot({ isBot }),
```

This is what I did to get it to work in [your project](https://github.com/intellix/ng-lazyload-image-issue-414):
1. Update ng-lazyload-image to version `7.0.1-0`
2. Remove:
```
"es2015": "fesm2015/ng-lazyload-image.js",
"esm2015": "esm2015/ng-lazyload-image.js",
"esm5": "esm5/ng-lazyload-image.js",
"fesm2015": "fesm2015/ng-lazyload-image.js",
"fesm5": "fesm5/ng-lazyload-image.js",
"module": "fesm5/ng-lazyload-image.js",
```
from `node_modules/ng-lazyload-image` to be sure what package webpack is using in the browser and server (`bundles/ng-lazyload-image.umd.js`).
3. Updates `node_modules/ng-lazyload-image/bundles/ng-lazyload-image.umd.js:~240` to `return Object.assign(ssrPreset, { isBot });`
4. Updated `src/app/app.module.ts` to:
```ts
export function isBot() {
  return true;
}
...
LazyLoadImageModule.forRoot({ isBot }),
```
5. Running `npm run build:ssr && npm run serve:ssr`
6. Verify by calling `curl http://localhost:4000/ | grep src`

**Comment by tjoskar** [comment]
I guess we need to:
1. Update `Object.assign(ssrPreset, { isBot})`
2. Update the readme with the information about how to write hooks for AOT.

**Comment by intellix** [comment]
Ah, you're right. I thought the original function worked cause the errors disappeared. I'll do a PR with the above ^

**Comment by tjoskar** [comment]
Included in 7.0.1
